pkg/server/image_pull.go

```diff
@@ -26,195 +26,223 @@ import (
+	// by someone else anytime, before/during/after we create the metadata. We should always
+	// check the actual state in containerd before using the image or returning status of the
+	// image.
 
-	image, err := c.imageStoreService.Get(ctx, resolvedImageName)
-	if err != nil {
-		return resolvedImageName, manifest, compressedSize,
-			fmt.Errorf("get failed for image:%q err: %v", resolvedImageName, err)
+	// Create corresponding image metadata.
+	newMeta := metadata.ImageMetadata{
+		ID:      imageID,
+		ChainID: chainID.String(),
+		Size:    size,
+		Config:  config,
 	}
-	p, err := content.ReadBlob(ctx, c.contentProvider, image.Target.Digest)
-	if err != nil {
-		return resolvedImageName, manifest, compressedSize,
-			fmt.Errorf("readblob failed for digest:%q err: %v", image.Target.Digest, err)
+	// Add the image reference used into repo tags. Note if the image is pulled with
+	// repo digest, it will also be added in to repo tags, which is fine.
+	updateImageMetadata(&newMeta, repoTag, repoDigest)
+	if err = c.imageMetadataStore.Create(newMeta); err != nil {
+		return nil, fmt.Errorf("failed to create image %q metadata: %v", imageID, err)
 	}
-	err = json.Unmarshal(p, &manifest)
-	if err != nil {
-		return resolvedImageName, manifest, compressedSize,
-			fmt.Errorf("unmarshal blob to manifest failed for digest:%q err: %v", image.Target.Digest, err)
-	}
-	size, err = image.Size(ctx, c.contentProvider)
-	if err != nil {
-		return resolvedImageName, manifest, compressedSize,
-			fmt.Errorf("size failed for image:%q %v", image.Target.Digest, err)
-	}
-	compressedSize = uint64(size)
-	return resolvedImageName, manifest, compressedSize, nil
+	return &runtime.PullImageResponse{ImageRef: imageID}, err
 }
 
-func (c *criContainerdService) pullImage(ctx context.Context, ref string) (imagespec.Descriptor, int64, error) {
-	var (
-		err               error
-		size              int64
-		desc              imagespec.Descriptor
-		resolvedImageName string
-		fetcher           remotes.Fetcher
-	)
-
-	// Resolve the image name; place that in the image store; then dispatch
-	// a handler for a sequence of handlers which: 1) fetch the object using a
-	// FetchHandler; and 3) recurse through any sub-layers via a ChildrenHandler
+// pullImage pulls image and returns image id (config digest) and digest.
+// TODO(random-liu): [P0] Wait for all downloadings to be done before return.
+func (c *criContainerdService) pullImage(ctx context.Context, ref string) (
+	imagedigest.Digest, imagedigest.Digest, error) {
+	// Resolve the image reference to get descriptor and fetcher.
 	resolver := docker.NewResolver()
-
-	resolvedImageName, desc, fetcher, err = resolver.Resolve(ctx, ref)
+	_, desc, fetcher, err := resolver.Resolve(ctx, ref)
 	if err != nil {
-		return desc, size, fmt.Errorf("failed to resolve ref %q: err: %v", ref, err)
+		return "", "", fmt.Errorf("failed to resolve ref %q: %v", ref, err)
```

Comment:
why change all these is there some rule that all errors in kube must return invalid return values for all params?